File: agent/src/FieldOps.Agent/Telemetry/TelemetryEnvelope.cs
using System.Text.Json;

namespace FieldOps.Agent.Telemetry;

internal enum TelemetryStatus
{
    Connecting,
    Ok,
    Degraded,
    Stale,
    Cached,
    Unavailable,
    Error,
}

internal sealed record TelemetrySource(
    string Id,
    string Type,
    string? Name = null,
    string? Version = null,
    JsonElement? Metadata = null);

internal sealed record TelemetryTimestamps(
    DateTimeOffset ObservedAt,
    DateTimeOffset ReceivedAt,
    DateTimeOffset? ExpiresAt = null);

internal sealed record TelemetryError(
    string Code,
    string Message,
    bool Retryable,
    JsonElement? Details = null,
    string? Cause = null);

/// <summary>
/// Transport-neutral telemetry snapshot. Ok and degraded states carry data;
/// stale and cached states may retain data; connecting and unavailable states
/// may omit data; error states carry structured error metadata.
/// </summary>
internal sealed record TelemetryEnvelope(
    TelemetryStatus Status,
    TelemetrySource Source,
    TelemetryTimestamps Timestamps,
    JsonElement? Data = null,
    TelemetryError? Error = null,
    JsonElement? Metadata = null)
{
    internal TelemetryEnvelope CreateOwnedCopy() => this with
    {
        Source = Source with { Metadata = CloneElement(Source.Metadata) },
        Data = CloneElement(Data),
        Error = Error is null ? null : Error with { Details = CloneElement(Error.Details) },
        Metadata = CloneElement(Metadata),
    };

    private static JsonElement? CloneElement(JsonElement? element)
    {
        if (element is not { } value || value.ValueKind == JsonValueKind.Undefined)
        {
            return element;
        }

        return value.Clone();
    }
}

File: src/telemetry/TelemetryEnvelope.ts
import type { TelemetryError } from './TelemetryError';
import type { TelemetryStatus } from './TelemetryStatus';

/** Identifies the producer and transport behind a telemetry reading. */
export interface TelemetrySource {
  readonly id: string;
  readonly type: string;
  readonly name?: string;
  readonly version?: string;
  readonly metadata?: Readonly<Record<string, unknown>>;
}

/** ISO-8601 timestamps describing the reading's lifecycle. */
export interface TelemetryTimestamps {
  readonly observedAt: string;
  readonly receivedAt: string;
  readonly expiresAt?: string;
}

interface TelemetryEnvelopeBase {
  readonly source: TelemetrySource;
  readonly timestamps: TelemetryTimestamps;
  readonly metadata?: Readonly<Record<string, unknown>>;
}

interface LiveTelemetryEnvelope<TPayload> extends TelemetryEnvelopeBase {
  readonly status: Extract<TelemetryStatus, 'ok' | 'degraded'>;
  readonly data: TPayload;
  readonly error?: never;
}

interface RetainedTelemetryEnvelope<TPayload> extends TelemetryEnvelopeBase {
  readonly status: Extract<TelemetryStatus, 'stale' | 'cached'>;
  readonly data?: TPayload;
  readonly error?: never;
}

interface PendingTelemetryEnvelope<TPayload> extends TelemetryEnvelopeBase {
  readonly status: Extract<TelemetryStatus, 'connecting' | 'unavailable'>;
  readonly data?: TPayload;
  readonly error?: never;
}

interface ErrorTelemetryEnvelope<TPayload> extends TelemetryEnvelopeBase {
  readonly status: Extract<TelemetryStatus, 'error'>;
  readonly data?: TPayload;
  readonly error: TelemetryError;
}

/** Transport-neutral telemetry snapshot narrowed by its lifecycle status. */
export type TelemetryEnvelope<TPayload = unknown> =
  | LiveTelemetryEnvelope<TPayload>
  | RetainedTelemetryEnvelope<TPayload>
  | PendingTelemetryEnvelope<TPayload>
  | ErrorTelemetryEnvelope<TPayload>;
